firedrake/preconditioners/patch.py

```diff
@@ -568,11 +571,41 @@ def select_entity(p, dm=None, exclude=None):
 class PlaneSmoother(object):
     @staticmethod
     def coords(dm, p):
-        coordsSection = dm.getCoordinateSection()
-        coordsDM = dm.getCoordinateDM()
-        dim = coordsDM.getDimension()
-        coordsVec = dm.getCoordinatesLocal()
-        return dm.getVecClosure(coordsSection, coordsVec, p).reshape(-1, dim).mean(axis=0)
+        mesh = dm.getAttr("__firedrake_mesh__")
+
+        ele = mesh.coordinates.function_space().ufl_element()
+        assert ele.family() in ["Lagrange", "Discontinuous Lagrange"]
+        if ele.family() == "Discontinuous Lagrange":
+            # Ah, we got a periodic mesh. We need to interpolate to CGk
+            # with access descriptor MAX to define a consistent opinion
+            # about where the vertices are.
+
+            coords = dm.getAttr("__firedrake_interpolated_coords__")
+            if coords is None:
+                CGkele = ele.reconstruct(family="Lagrange")
+                # Can't do
+                # CGk = FunctionSpace(mesh, CGkele)
+                # because equality fails between a mesh and a weakref.proxy.
+                CGk = FunctionSpace(mesh.coordinates.function_space().mesh(), CGkele)
+                coords = interpolate(mesh.coordinates, CGk, access=op2.MAX)
+                dm.setAttr("__firedrake_interpolated_coords__", coords)
```

Comment:
Why do you need to save this?

Reply:
The reason is that this creates a refcycle (which we're trying to avoid). You can hang it off the mesh if you like. Notice that this breaks if the mesh moves. Given that this setup only gets called once per solver object creation, I would rather just not cache this at all.

Reply:
The `coords` method is called once per plex point. I didn't want to interpolate the entire coordinate field once per plex point. I could store a weakref, but then someone would have to store a strong reference for it not to be garbage-collected?

As for the mesh moving --- I think everything in the `PlaneSmoother` construction has to be rebuilt then. It's probably safest for the user just to reconstruct the entire solver object?

Reply:
Can you move the creation of this to the `sort_entities` function then. Make the coordinates, pass them as an argument here and then delete them.

Reply:
Good point. I've moved it to `sort_entities`.